core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+                    exp_version, next_scheduled_check_time, state_name,
+                    feconf.TRAINING_JOB_STATUS_NEW, training_data,
+                    algorithm_version))
+            dummy_classifier_training_job.validate()
+
+            job_dicts_list.append({
+                'algorithm_id': algorithm_id,
+                'interaction_id': interaction_id,
+                'exp_id': exp_id,
+                'exp_version': exp_version,
+                'next_scheduled_check_time': next_scheduled_check_time,
+                'state_name': state_name,
+                'training_data': training_data,
+                'status': feconf.TRAINING_JOB_STATUS_NEW,
+                'algorithm_version': algorithm_version
+            })
+
+        job_ids = classifier_models.ClassifierTrainingJobModel.create_multi(
+            job_dicts_list)
+
+        for algorithm_id, job_id in python_utils.ZIP(
+                algorithm_ids_to_add, job_ids):
+            (
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                set_job_id_for_algorithm_id(algorithm_id, job_id))
+
+    if algorithm_ids_to_upgrade:
+        for algorithm_id in algorithm_ids_to_upgrade:
+            classifier_training_job = (
+                classifier_models.ClassifierTrainingJobModel.get_by_id(
+                    state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                    get_job_id_for_algorithm_id(algorithm_id)))
+            classifier_training_job.algorithm_version = (
+                algorithm_id_to_algorithm_version[algorithm_id])
+            classifier_training_job.next_scheduled_check_time = (
+                datetime.datetime.utcnow())
+            classifier_training_job.status = feconf.TRAINING_JOB_STATUS_NEW
+            classifier_training_job.put()
+
+    if algorithm_ids_to_remove:
+        for algorithm_id in algorithm_ids_to_remove:
+            delete_classifier_training_job(
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                get_job_id_for_algorithm_id(algorithm_id))
+            (
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                remove_algorithm_id(algorithm_id))
+
+    state_training_jobs_mapping_model = (
+        classifier_models.StateTrainingJobsMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    state_training_jobs_mapping.validate()
+    state_training_jobs_mapping_model.algorithm_ids_to_job_ids = (
+        state_training_jobs_mapping.algorithm_ids_to_job_ids.to_dict())
+    state_training_jobs_mapping_model.put()
+
+
+def get_classifier_training_job_maps(exp_id, exp_version, state_names):
+    """Gets the list of classifier training jobs mapped to algorithm id from
+    the exploration attributes.
 
     Args:
         exp_id: str. ID of the exploration.
         exp_version: int. The exploration version.
         state_names: list(str). The state names for which we retrieve the job.
```

Comment:
... the job mappings.

Reply:
@prasanna08 perhaps you missed this

Reply:
I actually addressed it but forgot to comment. Thanks.

Done.